CONTINUE_BUTTON_FEATURE

AnHardt opened the pull request:
CONTINUE_BUTTON defines a button for user interaction in case you have no display or panel.
To continue in a M0/M1, filament change, ...
When a BEEPER_PIN is defined beeps for one second.
Lights a LED while waiting if LED_PIN is defined.

`wait_for_user(millis_t until)` is the central function. It waits ms milliseconds
or until a button is clicked. The button is ether the click-button of a panel, or the newly introduced `CONTINUE_BUTTON`.

Example for the usage is M0/M1.

jbrazio commented:
I understand your idea.. but nowadays who runs a bot without Octoprint or a LCD ?..

AnHardt commented:
@jbrazio 
As far as i remember is was only a few days ago, when @Roxy-3D  suggested to misuse an endstop for this purpose.

jbrazio commented:
@AnHardt I deleted my comment because it came to my mind we recently had a discussion about this.

Roxy-3D commented:
Two questions:   

Should we consider debouncing this 'Continue Button' ?

```
+    #elif ENABLED(CONTINUE_BUTTON) && PIN_EXISTS(CONTINUE)
 +      while (PENDING(millis(), ms) && !continueButton) idle();
         while ( continueButton ) idle();     // This might be helpful???
 +    #endif
```

And the other questions...  Even though it is a 'mis-use' of an end-stop to do this, if I connect the continueButton up to an end-stop, I'm wondering if we want to do this after the continueButton is debounced:

```
endstops.hit_on_purpose(); // clear endstop hit flags
```

It may be acceptable to call that regardless of what the continueButton is connected to????

AnHardt commented on Marlin/Marlin_main.cpp:
```diff
@@ -8202,6 +8234,21 @@ void manage_inactivity(bool ignore_stepper_queue/*=false*/) {
+      continueCount++;
+    else if (continueCount > 0)
+      continueCount--;
+
```
@Roxy-3D 
For me this is a well working debouncer.

AnHardt replied:
Oops. Will change the comment later. (Changing code after the second beer is usually bad)

Done.

AnHardt commented:
If you suggest to call `endstops.hit_on_purpose()` here, then it means our current concept of disabling endstops is bad. I have an idea but have to sleep over it.

Instead (o)f undefining the pins we could

``` diff
     { // -direction
-        #if HAS_Y_MIN
+        #if HAS_Y_MIN && USE_YMIN_PLUG
          UPDATE_ENDSTOP(Y, MIN);
        #endif
      }
      else { // +direction
-        #if HAS_Y_MAX
+        #if HAS_Y_MAX && USE_YMAX_PLUG
          UPDATE_ENDSTOP(Y, MAX);
        #endif
      }
```

and so on.
Or something like this.

AnHardt commented:
Ideally we should use pin-change-interrupts for the endstops. That would save a bit of time in the stepper interrupt. (Example in grbl)  But it's not the time for that right now.

jbrazio commented:
Yes, in fact we should change the encoder to pin-change ISR also.. this is on our waiting list for sure.

Roxy-3D commented:
> If you suggest to call endstops.hit_on_purpose() here, then it means our current concept of disabling endstops is bad. I have an idea but have to sleep over it.

I'm not sure it means our current concept of disabling endstops is bad.   What we have here is a special case where (at least for the M600) the machine is not even moving and doesn't need any of its endstops.     But please do sleep on it!  Let's figure out a good solution!

AnHardt commented:
The unused endstop plugs are ideal for connecting other sensors/switches. GND and 5V is usually cho(s)e by. Not all boards have so ma(n)y usable plugs as a RAMPS. Didn't we have an issue today where someone tried to use an unused endstop?
However, simply undefini(n)g the pins is not good. Again we have the problem with the abstraction layer. Does `X_MIN_PIN` describe the location of the pin to find it on the board and a number where the processor can find it, or is that the pin we use for the x-min-endstop functionallity. That's not the same if we want to connect the continue button to it. 
I could imagine a two step mapping where `BOARD_PIN_X_MIN` maps to the number and `FUNCTION_PIN_X_MIN` maps to the `BOARD_PIN` - or, if unused, an other `FUNCTION_PIN` like `FUNKTION_PIN_CONTINUE_BUTTON` could be mapped to the free `BOARD_PIN`.
`BOARD_PIN`s could be named like `BOARD_PIN_AUX_2_3` on a RAMPS. That would help to find them and would not suggest a function. 

```
BOARD_PIN_STEPPER_SOCET_1_6
BOARD_PIN_MOSFET_1 //FAN
BOARD_PIN_POWER_MOSFET //extruder heater
BOARD_PIN_HIGH_POWER_HEATER_1 // bed heater
BOARD_PIN_ANALOG_3 // thermistor/analog sensor
BOARD_PIN_SERIAL_RX
BOARD_PIN_HW_SPI_MOSI
```

And then

``` cpp
#define FUNCTION_PIN_TEMP_ENVIRONMENT  BOARD_PIN_ANALOG_3
```

Roxy-3D commented:
> Does X_MIN_PIN describe the location of the pin to find it on the board and a number where the processor can find it, or is that the pin we use for the x-min-endstop functionality. That's not the same if we want to connect the continue button to it. 

Agreed!  I think I would lean towards Scott's thinking that if a board has some label silk screened on it, that is what we define the pin as.   But where opinions differ is I think it would be OK for us to:

```
#define CONTINUE_BUTTON X_MIN_PIN
```

What ever we decide...  We should be consistent and not have more than one philosophy in use.

esenapaj2 commented on Marlin/Configuration.h:
```diff
@@ -1117,6 +1117,15 @@ const bool Z_MIN_PROBE_ENDSTOP_INVERTING = false; // set to true to invert the l
 //
 //#define SAV_3DLCD
 
+// CUNTINUE_BUTTON defines a button for user interaction in case you have no display or panel.
```
Is CONTINUE_BUTTON better?

Roxy-3D replied:
Maybe it is better.   After all, that is the functionality being created.   But I typed that by mistake because in my head that is what I read 'CONTINUE_PIN' to mean.   We need to be consistent on the names.

thinkyhead commented:
Some boards have three plugs labeled as X-Stop, Y-Stop, Z-Stop, which we then map to `X_MIN_PIN`, `X_MAX_PIN`, etc., depending on the configured `USE_*_PLUG` options.

thinkyhead commented on Marlin/example_configurations/RepRapWorld/Megatronics/Configuration.h:
```diff
@@ -1123,6 +1123,15 @@ const bool Z_MIN_PROBE_ENDSTOP_INVERTING = false; // set to true to invert the l
+//#define CONTINUE_BUTTON_FEATURE
+//#define CONTINUE_PIN 35
+//#define BEEPER_PIN 37
+//#define LED_PIN 13
```
I don't think these pin overrides will work, currently. They might, but they will generate complaints from the compiler about multiple defines. Can the pin be explicitly pre-defined in the pins files? For example, to use a particular endstop by default (e.g., `#define CONTINUE_PIN X_MAX_PIN`) – or perhaps automatically use the first unallocated endstop pin, similar to how we assign the stepper pins for the Dual X, Y, and Z options…

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -8187,6 +8218,20 @@ void manage_inactivity(bool ignore_stepper_queue/*=false*/) {
     if (killCount >= KILL_DELAY) kill(PSTR(MSG_KILLED));
   #endif
 
+  #if ENABLED(CONTINUE_BUTTON_FEATURE) && CONTINUE_PIN != -1
```
`PIN_EXISTS(CONTINUE)` should be ok here too.

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -6833,12 +6868,12 @@ void process_next_command() {
 
     case 'M': switch (codenum) {
-      #if ENABLED(ULTIPANEL)
+      #if ENABLED(ULTIPANEL) || (ENABLED(ULTRA_LCD) && ENABLED(CONTINUE_BUTTON_FEATURE) && PIN_EXISTS(CONTINUE)) || (ENABLED(CONTINUE_BUTTON_FEATURE) && PIN_EXISTS(CONTINUE) && PIN_EXISTS(LED))
```
``` cpp
#if ENABLED(ULTIPANEL) \
  || (ENABLED(CONTINUE_BUTTON_FEATURE) && PIN_EXISTS(CONTINUE) \
       && (ENABLED(ULTRA_LCD) || PIN_EXISTS(LED)) \
     )
```

…or something else for `Conditionals.h`…

``` cpp
#define HAS_USER_INTERACTION ( ENABLED(ULTIPANEL) ||
    (ENABLED(CONTINUE_BUTTON_FEATURE) && PIN_EXISTS(CONTINUE) \
      && (ENABLED(ULTRA_LCD) || PIN_EXISTS(LED)) \
    ) \
  )
```

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -3794,7 +3807,38 @@ inline void gcode_G92() {
 }
 
-#if ENABLED(ULTIPANEL)
+#if ENABLED(ULTIPANEL) || (ENABLED(ULTRA_LCD) && ENABLED(CONTINUE_BUTTON_FEATURE) && PIN_EXISTS(CONTINUE)) || (ENABLED(CONTINUE_BUTTON_FEATURE) && PIN_EXISTS(CONTINUE) && PIN_EXISTS(LED))
```
``` cpp
#if HAS_USER_INTERACTION
```

…perhaps…

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -3794,7 +3807,38 @@ inline void gcode_G92() {
+      lcd_ignore_click(false);
+      while (PENDING(millis(), ms) && !lcd_clicked()) idle();
+	  lcd_ignore_click();
+    #elif ENABLED(CONTINUE_BUTTON_FEATURE) && PIN_EXISTS(CONTINUE)
```
If `PIN_EXISTS` is sanity-checked for `CONTINUE_BUTTON_FEATURE` then we don't have to check `PIN_EXISTS(CONTINUE)` here or elsewhere.

petrzjunior commented:
So guys, how does it look like? Is this feature getting to be finsihed? And any guide how to integrate in Filament Change?

thinkyhead commented:
I'm on the fence about this feature. If it is rebased I will probably merge it anyway. @petrzjunior The only change you'll need is to call the new `wait_for_user` function in `M600` instead of using `while (!LCD_CLICKED) idle();`.

AnHardt committed:
CONTINUE_BUTTON_FEATURE

CONTINUE_BUTTON_FEATURE defines a button for user interaction in case you have no display or panel.
To continue in a M0/M1, filament change, ...
When a BEEPER_PIN is defined beeps for one second.
Lights a LED while waiting if LED_PIN is defined.

`wait_for_user(millis_t until)` is the central function. It waits ms milliseconds
or until a button is clicked. The button is ether the click-button of a panel, or the newly introduced `CONTINUE_BUTTON`.

Example for the usage is M0/M1.

thinkyhead commented:
The other day I didn't have a controller connected, so I had `M0`/`M1` set a flag "`wait_for_user`" and wait in a loop. Then I had `M108` clear the flag. I could have had it use an endstop, but `M108` was fine for the purpose.
